### PR TITLE
New `DepthImage` archetype

### PR DESCRIPTION
**Pull request by emilk**
### What
* Part of https://github.com/rerun-io/rerun/issues/6386

This changes `DepthImage` from a tensor to a byte-blob + resolution + data-type.
It adds new code for hovering images and uploading images to GPU, both general enough to also work for color images and segmentation images. This means those PRs will be a lot smaller.

Since we still need to support color and segmentation images encoded as tensors, the code is quite complex, with duplicated paths here and there. It will all get cleaned up in later PRs.

I'm reusing a lot of stuff from tensors, like `TensorMeaning`, `TensorStats` `TensorElement`, but these will be renamed and improved before https://github.com/rerun-io/rerun/issues/6386 is done.

### TODO
* [x] Update `depth_image_ext.cpp`
* [x] Update `depth_image_ext.py`
* [x] Fix hovering depth images in 3D view

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using examples from latest `main` build: [rerun.io/viewer](https://rerun.io/viewer/pr/6915?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [rerun.io/viewer](https://rerun.io/viewer/pr/6915?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [x] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!
* [x] If have noted any breaking changes to the log API in `CHANGELOG.md` and the migration guide
* [x] Test a few local examples

- [PR Build Summary](https://build.rerun.io/pr/6915)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

### CODE REVIEW

**Comment by emilk** [comment]
@rerun-bot full-check

**Comment by emilk** [comment]
@rerun-bot full-check

File: crates/store/re_types_core/src/arrow_buffer.rs [hunk 1]
```diff
@@ -56,6 +56,28 @@ impl<T> ArrowBuffer<T> {
...
+impl<T: bytemuck::Pod> ArrowBuffer<T> {
+    /// Cast POD (plain-old-data) types to another POD type.
+    ///
+    /// For instance: cast a buffer of `u8` to a buffer of `f32`.
+    #[inline]
+    pub fn cast_pod<Target: bytemuck::Pod>(
+        &self,
+    ) -> Result<ArrowBuffer<Target>, bytemuck::PodCastError> {
+        let target_slice: &[Target] = bytemuck::try_cast_slice(self.as_slice())?;
+        Ok(ArrowBuffer::from(target_slice.to_vec()))
```
Comment (emilk) [inline crates/store/re_types_core/src/arrow_buffer.rs:68]: Maybe there is a zero-copy version of this, or else create a TODO/issue for it

File: docs/snippets/all/archetypes/depth_image_3d.cpp [hunk 2]
```diff
@@ -31,7 +31,7 @@ int main() {
 
     rec.log(
         "world/camera/depth",
-        rerun::DepthImage({HEIGHT, WIDTH}, data)
+        rerun::DepthImage({WIDTH, HEIGHT}, data)
```
Comment (emilk) [inline docs/snippets/all/archetypes/depth_image_3d.cpp:34]: TODO: change parameter orders here to force users to rethink their log code and read migration guide.

Also, "data, length" is our most common parameter order

File: rerun_cpp/src/rerun/archetypes/depth_image_ext.cpp [hunk 3]
```diff
@@ -1,75 +1,47 @@
...
+        : DepthImage{resolution_, get_data_type(pixels), reinterpret_cast<const uint8_t*>(pixels)} {
+    }
 
-        // We want to change the dimension names if they are not specified.
-        // But rerun collections are strictly immutable, so create a new one if necessary.
-        bool overwrite_height = !shape[0].name.has_value();
-        bool overwrite_width = !shape[1].name.has_value();
+    template <typename TElement>
+    DepthImage(components::Resolution2D resolution_, std::vector<TElement> pixels)
+        : DepthImage{resolution_, Collection<TElement>::take_ownership(std::move(pixels))} {}
```
Comment (emilk) [inline rerun_cpp/src/rerun/archetypes/depth_image_ext.cpp:18]: should be covered by the `Collection` overload?

File: rerun_cpp/src/rerun/archetypes/depth_image_ext.cpp [hunk 4]
```diff
@@ -1,75 +1,47 @@
...
 
-            if (overwrite_height) {
-                new_shape[0].name = "height";
-            }
-            if (overwrite_width) {
-                new_shape[1].name = "width";
-            }
+    /// New depth image from an `ChannelDataType` and a pointer.
+    ///
+    /// The length of the data should be `W * H * data_type.size`
```
Comment (emilk) [inline rerun_cpp/src/rerun/archetypes/depth_image_ext.cpp:26]: document that the data is borrows

File: rerun_cpp/src/rerun/archetypes/depth_image_ext.cpp [hunk 5]
```diff
@@ -1,75 +1,47 @@
...
-
-    DepthImage::DepthImage(components::TensorData data_) : data(std::move(data_)) {
-        auto& shape = data.data.shape;
-        if (shape.size() != 2) {
-            std::stringstream ss;
-            ss << "Expected 2-dimensional tensor, got " << shape.size() << " dimensions.";
-            Error(ErrorCode::InvalidTensorDimension, ss.str()).handle();
-            return;
-        }
+    DepthImage(components::Resolution2D resolution_, const TElement* pixels)
```
Comment (emilk) [inline rerun_cpp/src/rerun/archetypes/depth_image_ext.cpp:12]: Document that the data is borrows

File: rerun_cpp/src/rerun/archetypes/depth_image_ext.cpp (hunk 3 above)
Comment (emilk) [reply rerun_cpp/src/rerun/archetypes/depth_image_ext.cpp]: nope, this is needed

**Review by Wumpf — APPROVED** [review]
so much _stuff_ :O

File: crates/store/re_types/src/archetypes/image_encoded_ext.rs [hunk 6]
```diff
@@ -20,6 +20,7 @@ impl ImageEncoded {
     ///
     /// [`Self::media_type`] will be guessed from the bytes.
     pub fn from_file_contents(bytes: Vec<u8>) -> Self {
+        #[allow(clippy::unnecessary_struct_initialization)]
```
[the remark this answers is not in the document]
Comment (emilk) [reply crates/store/re_types/src/archetypes/image_encoded_ext.rs:23]: Needed with some feature flags (or rather, lack of the `image` feature flag)

File: crates/viewer/re_viewer_context/src/image_info.rs [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    pub fn bits_per_texel(&self) -> usize {
+        // TODO(#6386): use `PixelFormat`
+        self.data_type.bits() * self.num_channels()
+    }
+
+    /// Get the value of the element at the given index.
+    ///
+    /// Return `None` if out-of-bounds.
+    #[inline]
+    pub fn get_xyc(&self, x: u32, y: u32, channel: u32) -> Option<TensorElement> {
```
[the remark this answers is not in the document]
Comment (emilk) [reply crates/viewer/re_viewer_context/src/image_info.rs:67]: I just want to be explicit with the parameter order, since for tensors it is usually `yxc`

**Comment by emilk** [comment]
@rerun-bot full-check

**Pull request merged**